FileSystem updates and test modernization to fix intermittent test failures and concurrency issues.

ianhays opened the pull request:
- System.IO.UnixFileSystem has an issue with the "overwrite" being interpreted backwards. I simply reversed the conditional statement to be correct. There were previously no tests to cover this possibility. Resolves #2276
- Updated over a dozen test files to use the modern Xunit syntax.
- Solved file-in-use concurrency issues in the altered tests. Resolves #1313, #2227
- Tests will no longer intermittently fail.
- Slightly improved code coverage for the FileSystem files whose tests I altered (Directory, DirectoryInfo, File, FileInfo)
- Most tests altered now inherit from similar tests to consolidate the testing robustness into as few locations as possible. For example, Directory_GetFiles_str now inherits from Directory_GetFileSystemEntries_str since the tests for Directory_GetFiles_str were the exact same ones as in Directory_GetFileSystemEntries_str anyways.
- All altered tests now use the System.IO.FileSystem.Tests namespace
- Renamed the File_WriteAll_all class name/file name to File_WriteAllText.
- I tested my changes on Windows overnight to verify that there is no more intermittent test failures. I didn't have to modify all of the files listed in #2227 to get them to pass; this concerns me but I found no way to replicate the errors I was receiving before. My guess would be that the tests in the files I didn't alter (e.g. File_WriteAll and File_ReadAll) used the same names as the tests I did alter but now that the tests I altered are changed they no longer share test file names and no longer fail as a result.

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetDirectories_str.cs:
```diff
@@ -1,377 +1,54 @@
+                    string linkPath = Path.Combine(containingFolder.Path, Path.GetRandomFileName());
+                    Assert.Equal(0, symlink(targetDir.Path, linkPath));
+                    Assert.True(Directory.Exists(linkPath));
+                    Assert.Equal(1, Directory.GetDirectories(containingFolder.Path).Count());
```
These Directory.* calls should now be using the instance Get* methods on this test class, right?

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str.cs:
```diff
@@ -1,221 +1,219 @@
-    {
-        Assert.Throws<ArgumentException>(() => GetEntries(String.Empty));
-    }
+        public static string[] WhitespaceVariations = new string[] { "         ", " ", "\n", ">", "<", "\t" };
```
It's strange that '<' and '>' are included in something called "Whitespace".

ianhays replied:
Agreed. It was originally just used for whitespace, hence the name, but I later added the arrows and neglected to change it. I'll change it to something like "WindowsInvalidUnixValid" that tells you more about its purpose.

stephentoub commented on src/System.IO.FileSystem/tests/File/Create_str.cs:
```diff
@@ -0,0 +1,231 @@
+            using (Create(string.Format(" {0}", testFile1)))
+            using (Create(testFile3))
+            {
+                Assert.True(File.Exists(testFile1));
```
Isn't this going to fail on Unix?

ianhays replied:
They pass on Ubuntu.

stephentoub replied:
Not for me it doesn't.  And that makes sense, because the string being passed to create starts with a space, which means it's not rooted, which means it's going to be appended to the current directory.  So if you were trying to create /home/ianhays/corefx/testfile1, and you're running this from /home/ianhays/corefx, it's actually going to try to create "/home/ianhays/corefx/ /home/ianhays/corefx/testfile1", right?

ianhays replied:
Alright, I figured it out. My Unix build was outdated and didn't have the changes you made to Path. In the build I have, Path.Combine will just remove the extra whitespace from the path for both files, making the paths absolute and removing the directory with the one-space name.

There are other changes in this PR that will be altered by me updating my Unix build. This also explains why I reopened that issue - the changes weren't reflected yet to me. I'm going to change the way I test on Unix so that this doesn't happen again.

#2205 can likely be closed.

stephentoub replied:
:smile:

ianhays replied:
Okay, now everything is failing properly :) 

This method still has value to test that extra whitespace in a filename is removed before file creation, but it should be marked as Windows only and a similar Unix version should be added. I'm also planning now to add a similar test to GetFileSystemEntries, as I don't think I'm currently testing it there either.

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs:
```diff
@@ -1,221 +1,371 @@
-    }
+        [Fact]
+        [PlatformSpecific(PlatformID.Windows)]
+        public void WindowsSearchPatternQuestionMarks()
```
I just tried running the tests on Windows, and this fails for me repeatedly due to paths being too long.

ianhays replied:
Failed the build system too. I thought I'd removed all the deeply nested test directories, but I appear to have missed this one. Thanks.

stephentoub commented on src/System.IO.FileSystem/tests/FileInfo/get_LastWriteTime.cs:
```diff
@@ -1,131 +1,45 @@
+            DateTime afterWrite = DateTime.Now;
+            fileInfo.Refresh();
+            DateTime lastWriteTime = fileInfo.LastWriteTime;
+            Assert.InRange<long>(fileInfo.LastWriteTime.Ticks, beforeWrite.Ticks, afterWrite.Ticks);
```
This could be problematic on Linux (not sure about OSX).  Our current implementation uses utime, with a resolution of one second I believe.  If that's true, we could easily have a case where, for example, the last write time is at 1:23:45, the before time is 1:23:45.67 and the after time is 1:23:45.89... everything is correct, but the InRange will fail.

If I'm right that this is problematic, I'd suggest loosening the test a bit to allow some buffer.  Separately, you might open an issue and/or look to address the implementation to see about providing more accuracy, e.g. using utimes instead of utime, but we might want to wait for #2137 before doing that, as if memory serves it involves a structure that'll be very fragile to declare in managed code.

ianhays replied:
I'll add a one-second negative to the before time to be safe.

stephentoub commented on src/System.IO.FileSystem/tests/File/Copy_str_str.cs:
```diff
@@ -1,439 +1,150 @@
                 {
-                    iCountErrors++;
-                    printerr("Error_t948v! Source file gone");
+                    Copy(testFile, whitespace);
```
Is this not problematic, trying to copy the file that's still open from the File.Create in the using that wraps this?

stephentoub replied:
Also, this is copying to a file in the current directory named for whatever is in whitespace... that's going to leave behind temporary files, and likely cause problems when this method runs as part of one of the other derived types and tries to copy to a file that already exists.

stephentoub commented on src/System.IO.FileSystem/tests/Directory/Move_str_str.cs:
```diff
@@ -1,430 +1,281 @@
+            Move(testDirToMove, testDir + "         ");
+            Move(testDir + "         ", testDir + "\n");
+            Move(testDir + "\n", testDir + "\t");
+            Move(testDir + "\t", testDir + "");
```
Seems like this should fail because we're trying to move the subdirectory to overwrite the parent (the string being appended is the empty string).

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str.cs:
```diff
@@ -1,221 +1,219 @@
+
+        [Fact]
+        [PlatformSpecific(PlatformID.AnyUnix)]
+        [ActiveIssue(2205)]
```
Why is this marked as an ActiveIssue?

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str.cs:
```diff
@@ -1,221 +1,219 @@
+
+        [Fact]
+        [PlatformSpecific(PlatformID.AnyUnix)]
+        [ActiveIssue(2205)]
```
Why is this marked as an ActiveIssue?

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs:
```diff
@@ -1,221 +1,371 @@
+
+        [Fact]
+        [PlatformSpecific(PlatformID.AnyUnix)]
+        [ActiveIssue(2205)]
```
Why is this marked as an ActiveIssue?

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs:
```diff
@@ -1,221 +1,371 @@
+
+        [Fact]
+        [PlatformSpecific(PlatformID.AnyUnix)]
+        [ActiveIssue(2205)]
```
Why is this marked as an ActiveIssue?

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str.cs:
```diff
@@ -1,221 +1,219 @@
-            Assert.Contains(Path.Combine(testDirPath, " "), results);
-            Assert.Contains(Path.Combine(testDirPath, "          "), results);
-            Assert.Contains(Path.Combine(testDirPath, "\n"), results);
+            string strTempDir = Path.Combine("dls;d", "442349-0", "v443094(*)(+*$#$*") + new string(Path.DirectorySeparatorChar, 3);
```
Should there be a corresponding AnyUnix test demonstrating that these don't fail?

ianhays replied:
I actually think I'll remove this test and replace it with something that calls Path.GetInvalidFileNameChars so it can be more xplat friendly. I copied this one from the old code before adding new tests that mostly eclipsed its original purpose.

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str.cs:
```diff
@@ -1,221 +1,219 @@
+        public void WindowsWhitespacePath()
+        {
+            foreach (string whitespace in WhitespaceVariations)
+                Assert.Throws<ArgumentException>(() => GetEntries(whitespace));
```
This could be done using Assert.All, e.g.
```C#
Assert.All(WhitespaceVariations, whitespace =>
    Assert.Throws<ArgumentException>(() => GetEntries(whitespace)));
```

ianhays replied:
Oh cool! I've never seen that before. I'll use that.

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str.cs:
```diff
@@ -1,221 +1,219 @@
+            {
+                DirectoryInfo testDir = Directory.CreateDirectory(GetTestFilePath());
+                foreach (string whitespace in WhitespaceVariations)
+                    File.Create(testDir.FullName + Path.DirectorySeparatorChar + whitespace).Dispose();
```
Path.Combine?

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str.cs:
```diff
@@ -1,221 +1,219 @@
+
+                string[] results = GetEntries(testDir.FullName);
+                foreach (string whitespace in WhitespaceVariations)
+                    Assert.Contains(testDir.FullName + Path.DirectorySeparatorChar + whitespace, results);
```
Path.Combine? (I'll stop asking :)

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs:
```diff
@@ -1,221 +1,371 @@
+            // string passed. But we should return 0 items.
+            Assert.Empty(GetEntries(TestDirectory, string.Empty));
+        }
+        public void SearchPatternValid()
```
Missing [Fact]?

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs:
```diff
@@ -1,221 +1,371 @@
-        using (File.Create(testFile1))
-        using (File.Create(testFile2))
+        [Fact]
+        [ActiveIssue(645)]
```
Rather than this being an ActiveIssue, should this instead verify that an exception is thrown for a long path?  Then when long path support is added, the test should start failing and it can be fixed then?

ianhays replied:
If that's the preferred way, then sure :) I'll just comment out the test I have and mark it with a TODO referencing 645, then add a PathTooLongException check.
.

stephentoub commented on src/System.IO.FileSystem/tests/Directory/Move_str_str.cs:
```diff
@@ -1,430 +1,281 @@
-            }
-            dir2.Delete(true);
+        [Fact]
+        [ActiveIssue(645)]
```
Same question about long path; I'll stop asing this one, too :)

stephentoub commented on src/System.IO.FileSystem/tests/DirectoryInfo/Refresh.cs:
```diff
@@ -1,126 +1,48 @@
-        if (iCountErrors != 0)
+
+        [Fact]
+        public void AttributeChange()
```
ReadOnly is one of or maybe the only attribute that'll roundtrip like this on Unix.  Do we have a similar test somewhere that tests the other attributes on Windows?

ianhays replied:
The closest thing is the tests in "set_attributes", although those don't call Refresh() or create a new FileAttribute(), they just set and check and move on to the next attribute.

stephentoub commented on src/System.IO.FileSystem/tests/File/Copy_str_str.cs:
```diff
@@ -1,439 +1,150 @@
             {
-                iCountErrors++;
-                printerr("Error_109uc! Incorrect exception thrown, exc==" + exc.ToString());
+                Assert.Throws<IOException>(() => Copy(testFile, Directory.GetCurrentDirectory()));
```
I don't think this is testing what it's meant to be testing.  It's going to pass, but only because the resulting exception from trying to copy an open file is an IOException just as is the exception for copying onto a directory.

stephentoub commented on src/System.IO.FileSystem/tests/File/Copy_str_str.cs:
```diff
@@ -1,439 +1,150 @@
             {
-                iCountErrors++;
-                printerr("Error_209us! Incorrect exception thrown, exc==" + exc.ToString());
+                Assert.Throws<IOException>(() => Copy(testFile, testFile));
```
Same here.

stephentoub commented on src/System.IO.FileSystem/tests/File/Create_str.cs:
```diff
@@ -0,0 +1,231 @@
+            using (Create(testDir.FullName + Path.DirectorySeparatorChar + " "))
+            using (Create(testDir.FullName + Path.DirectorySeparatorChar + "\n"))
+            using (Create(testDir.FullName + Path.DirectorySeparatorChar + ">"))
+            using (Create(testDir.FullName + Path.DirectorySeparatorChar + ">"))
```
Is this a typo?  I'd expect this to fail, not because of the FileSystem implementation, but because this same directory is being created twice and then below we're checking for one that doesn't exist.

ianhays replied:
Ah, yeah, typo.

stephentoub commented on src/System.IO.FileSystem/tests/File/WriteAllText.cs:
```diff
@@ -0,0 +1,532 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Was this file just renamed?  It doesn't appear to have been rewritten.

ianhays replied:
Yes

stephentoub commented:
Overall, very nice job... things are looking so much better!

There are some issues that need to be addressed, but definitely headed in the right direction.

ianhays commented:
Updated in response to feedback. A few things:
-Removed ActiveIssue references, particularly about long paths and unix whitespace
-Before Path.Combine worked with unix whitespace-only names, I wrote my tests using string concatenation. I have switched those instances to instead use Path.Combine where I could find them.
-Tested on Unix far more extensively and fixed a few bugs mentioned in the comments
-I didn't change existing loops to Assert.All, but I will write write my future loops in that way.
-The commit history includes every time I merged my enlistment from the dotnet upstream. Do I need to squash those again, or will the PR merge get rid of them for me?

@stephentoub Thanks, and have a great weekend!

davidsh commented:
/cc @ellismg 

The PR merge will not squash/merge those commits.  Matt Ellis has some git techniques to help you roll back and rebase your PR so that you can have a clean commit on the latest master.

ellismg commented:
@ianhays To squash you can do the following:

(Make sure all your work is committed and you are on master)

```
> git reset --soft 1764a71
> git commit
```

Write a new commit message and save. 

At that point if you look at your history you should have just one commit sitting on top of master, you can then force push your history. If things end up looking fishy you can get back to where you were before with `git reset --hard e4de1f3`.  If you run into problems along the way, let me know and I can help you get unstuck.

In the future, it would be preferable for you to make a local branch in your clone instead of working off of master and to rebase instead of merge when you want to bring in new changes from master.

ianhays commented:
@davidsh @ellismg Thanks. That was much simpler than what I ended up doing to squash in my last PR.

Will do!

stephentoub commented:
Four failures on Windows:
```
System.IO.FileSystem.Tests.Directory_EnumerateFileSystemEntries_str_str_so.SearchPatternIgnoreSubDirectories [FAIL]
        System.IO.PathTooLongException : The specified file name or path is too long, or a component of the specified path is too long.
        Stack Trace:
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.IO.FileSystem\src\System\IO\Win32FileSystem.cs(113,0): at System.IO.Win32FileSystem.CreateDirectory(String fullPath)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.IO.FileSystem\src\System\IO\Directory.cs(43,0): at System.IO.Directory.CreateDirectory(String path)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.IO.FileSystem\tests\Directory\GetFileSystemEntries_str_str.cs(151,0): at System.IO.FileSystem.Tests.Directory_GetFileSystemEntries_str_str.SearchPatternIgnoreSubDirectories()
     System.IO.FileSystem.Tests.Directory_EnumerateFileSystemEntries_str_str_so.SearchPatternIncludeSubDirectories [FAIL]
        System.IO.PathTooLongException : The specified file name or path is too long, or a component of the specified path is too long.
        Stack Trace:
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.IO.FileSystem\src\System\IO\Win32FileSystem.cs(113,0): at System.IO.Win32FileSystem.CreateDirectory(String fullPath)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.IO.FileSystem\src\System\IO\Directory.cs(43,0): at System.IO.Directory.CreateDirectory(String path)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.IO.FileSystem\tests\Directory\GetFileSystemEntries_str_str_so.cs(60,0): at System.IO.FileSystem.Tests.Directory_GetFileSystemEntries_str_str_so.SearchPatternIncludeSubDirectories()
     System.IO.FileSystem.Tests.DirectoryInfo_EnumerateFileSystemInfos_str_str_so.SearchPatternIgnoreSubDirectories [FAIL]
        System.IO.PathTooLongException : The specified file name or path is too long, or a component of the specified path is too long.
        Stack Trace:
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.IO.FileSystem\src\System\IO\Win32FileSystem.cs(113,0): at System.IO.Win32FileSystem.CreateDirectory(String fullPath)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.IO.FileSystem\src\System\IO\Directory.cs(43,0): at System.IO.Directory.CreateDirectory(String path)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.IO.FileSystem\tests\Directory\GetFileSystemEntries_str_str.cs(151,0): at System.IO.FileSystem.Tests.Directory_GetFileSystemEntries_str_str.SearchPatternIgnoreSubDirectories()
     System.IO.FileSystem.Tests.DirectoryInfo_EnumerateFileSystemInfos_str_str_so.SearchPatternIncludeSubDirectories [FAIL]
        System.IO.PathTooLongException : The specified file name or path is too long, or a component of the specified path is too long.
        Stack Trace:
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.IO.FileSystem\src\System\IO\Win32FileSystem.cs(113,0): at System.IO.Win32FileSystem.CreateDirectory(String fullPath)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.IO.FileSystem\src\System\IO\Directory.cs(43,0): at System.IO.Directory.CreateDirectory(String path)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.IO.FileSystem\tests\Directory\GetFileSystemEntries_str_str_so.cs(60,0): at System.IO.FileSystem.Tests.Directory_GetFileSystemEntries_str_str_so.SearchPatternIncludeSubDirectories()
```

ianhays commented:
Alright, that's twice now that my build fails because of unexpected PathTooLongExceptions. I'm going to start running my tests from a directory with the same full path length as the one in Jenkins so these errors don't unexpectedly pop up again!

Tests are passing now; need to squash the new commit.

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs:
```diff
@@ -1,221 +1,378 @@
+            Assert.Empty(GetEntries(TestDirectory, "           "));
+            Assert.Empty(GetEntries(TestDirectory, "\n"));
+            Assert.Empty(GetEntries(TestDirectory, " "));
+            Assert.Empty(GetEntries(TestDirectory, "\t"));
```
Minor: this could be done with a [Theory].

ianhays replied:
I actually just learned about the [Theory] tag about 2 minutes ago from looking at your MemoryMapped PR! Definitely going to be using it in the next tests I write; there's several occasions in this code where I could have made use of it.

stephentoub commented on src/System.IO.FileSystem/tests/DirectoryInfo/EnumerableAPIs.cs:
```diff
@@ -1,1047 +1,138 @@
-// Copyright (c) Microsoft. All rights reserved.
+// Copyright (c) Microsoft. All rights reserved.DirectoryInfo_EnumerateFiles_Str
```
Typo in the header

ianhays replied:
That's... odd.

stephentoub commented:
Skimmed the revisions.  LGTM.  Thanks for revising these.

ianhays commented:
@stephentoub thanks! Pushed that typo change :) Will merge after build success.
